rfcs/RFC-Upload-From-Dropbox.md

```diff
@@ -0,0 +1,367 @@
+upload to S3. For calculating the checksum while streaming, [checksumming_io](https://github.com/HumanCellAtlas/checksumming_io)
+can be used.
+
+Once the upload is complete, the download job will be removed from the queue. If the final hash does not match, then
+the upload process will be restarted. The retry logic will be an exponential backoff, with a max retry of 5 times.
+
+While streaming the file, the compute resource will periodically check if the upload has been canceled using the upload
+table. If the upload status is cancel pending, the compute resource will stop the stream, delete any data that has been uploaded,
+mark the job as complete in the queue, and set the status of the upload to cancelled.
+
+#### 7. Retry
+
+If the upload fails for any reason, the compute resource will mark the status of the upload as waiting, and return the upload
+job to the queue.
+
+#### 8. Failed Uploads
+
+If an upload fails enough times, it will be removed from the download queue. The upload status will change to failed.
+
+### Components
+
+This section describes the purpose of components that make up the upload design.
+
+#### Data Portal Browser App
+
+The DP Browser App is the frontend of the system. It is the primary method users have to interact with the Data Portal.
+It will be used to authenticate the user and interact with the DP Backend to upload files.
+
+#### Data Portal Backend
+
+The DP Backend is responsible for handling API requests. It will enqueue shared links for downloading, and update the
+upload status. This additional functionality can be built into the existing DP Backend Lambda.
+
+The backend will also be responsible for performing a simple validation of the file being uploaded. The validation is
+performed during the initial request to upload. The validation consists of verifying the extension of the file is a [supported type](https://github.com/chanzuckerberg/corpora-data-portal/blob/main/backend/schema/corpora_schema.md#implementations).
+
+#### Upload Queue
+
+The Upload Queue is an [AWS SQS](https://aws.amazon.com/sqs/) that will track pending upload jobs.
+The upload jobs are used by the upload service to know where to download the file from before uploading them to S3.
+For the first implementation, the job sleep time will be 15min. Eventually, the job sleep time will depend on the size
+of the file being uploaded.
+
+Cloudwatch events can be used to monitor the SQS length. Additional compute resource can be spawned in response to an
+increase in jobs in the queue.
+
+##### Upload Job Entry
+
+These are the fields that will be in the upload job placed in the queue:
+
+- Link - the link to file to download. Use to retrieve the file from DropBox.
+- submission_id - identifies the submission. Used to determine the storage location.
+- dataset_id - identifies the dataset. Used to determine the storage location.
+- file_name - the name of the file being downloaded.
```

Comment:
How do you determine this from the Dropbox link?

Reply:
There is a `filename` field in the header, there is also a `content-length` field to give the size of the file.